File: src/array/buy_sell.cpp
#include <vector>
#include <iostream>
#include <algorithm>
#include <unordered_map>

using namespace std;

int bruteForce(vector<int> & prices) {
	int maxprofit = 0;
	for (int i = 0; i < prices.size(); i++) {
		for (int j = i + 1; j < prices.size(); j++) {
			int profit = prices[j] - prices[i];
			if (profit > maxprofit) {
				maxprofit = profit;
			}
		}
	}
	return maxprofit;
}

int maxProfit(vector<int> & prices) {
	int minprice = INT_MAX;
	int maxprofit = 0;
	for (int i = 0; i < prices.size(); i++) {
		if (prices[i] < minprice) {
			prices[i] = minprice;
		} else if (prices[i] - minprice > maxprofit) {
			maxprofit = (prices[i] - minprice);
		}
	}
	return maxprofit;
}

int main() {
	vector<int> arr = {1, 3, 1, 4, 5, 9, 2, 10, 7}; // max profit is 9
	cout << bruteForce(arr) << endl;
}

File: src/array/buy_sell.rs
use i32::{max, min};

fn max_profit(prices: Vec<i32>) -> i32 {
    let mut min_price = prices[0];
    let mut max_diff = 0;

    for i in 1..prices.len() {
        min_price = min(min_price, prices[i]);
        max_diff = max(max_diff, prices[i] - min_price);
    }

    return max_diff;
}

fn main() {
    let x: i32 = max_profit(vec![1, 1]);
    let y: i32 = max_profit(vec![2, 2, 3, 1, 4, 7]);

    println!("x: {}, y: {}", x, y);
}
